### PR TITLE
Refactor is_a

### PR DESCRIPTION
**Pull request by samsonasik**
Fixes https://github.com/rectorphp/rector/issues/5906

### CODE REVIEW

**Comment by samsonasik** [comment]
@ondrejmirtes this is_a change https://github.com/rectorphp/rector/pull/5907/commits/5ac198c7d332258ce53c7a19c9a1a452cb3b6102 seems cause error in test:

```bash
1) Rector\Tests\NetteToSymfony\Rector\Interface_\DeleteFactoryInterfaceRector\DeleteFactoryInterfaceFileSystemRectorTest::test with data set #0 (Symplify\SmartFileSystem\SmartFileInfo Object (...))
61
Failed asserting that false is true.
```

@TomasVotruba PHPStan notice seems unrelated as i tried in local and got same error with or without this change:

```bash
vendor/bin/phpstan analyse --ansi --error-format symplify packages/BetterPhpDocParser/ValueObject/PhpDocNode/JMS/SerializerTypeTagValueNode.php
Note: Using configuration file /Users/samsonasik/www/rector/phpstan.neon.
PHP Fatal error:  Class Rector\BetterPhpDocParser\ValueObject\PhpDocNode\JMS\SerializerTypeTagValueNode contains 1 abstract method and must therefore be declared abstract or implement the remaining methods (PHPStan\PhpDocParser\Ast\Node::hasAttribute) in /Users/samsonasik/www/rector/packages/BetterPhpDocParser/ValueObject/PhpDocNode/JMS/SerializerTypeTagValueNode.php on line 13
Fatal error: Class Rector\BetterPhpDocParser\ValueObject\PhpDocNode\JMS\SerializerTypeTagValueNode contains 1 abstract method and must therefore be declared abstract or implement the remaining methods (PHPStan\PhpDocParser\Ast\Node::hasAttribute) in /Users/samsonasik/www/rector/packages/BetterPhpDocParser/ValueObject/PhpDocNode/JMS/SerializerTypeTagValueNode.php on line 13
```

**Comment by TomasVotruba** [comment]
@samsonasik I need to merge this https://github.com/rectorphp/rector/pull/5841 first

**Comment by samsonasik** [comment]
@TomasVotruba thank you 👍

**Comment by samsonasik** [comment]
@ondrejmirtes I got the patch about failing test case, the **UnionType** need to check against **ShortenedObjectType** to make **$returnType** re-assigned with new ObjectType($returnType->getFullyQualifiedName()) https://github.com/rectorphp/rector/pull/5907/commits/f4ee4c45d1989d67687df0f5325f34f0cd6dad32

**Comment by samsonasik** [comment]
test error start from change https://github.com/rectorphp/rector/pull/5907/commits/79f8698e6a1ca8cc291ce9e467df663b711d435b

```diff
-        return is_a($currentType->getClassName(), $newType->getClassName(), true);
+        return $newType->isSuperTypeOf($currentType)->yes();
```

**Comment by samsonasik** [comment]
try check one by one, it seems caused by https://github.com/rectorphp/rector/pull/5907/commits/2bc0a986cc36c01edb80816e0e5137aa9e87bf4b

**Comment by samsonasik** [comment]
somehow, it fixed with return early on `yes()` check instead of `no()` and continue: 

```diff
-                if (! is_a($type->getClassName(), $countableObjectType->getClassName(), true)) {
-                     continue;
+                 if ($countableObjectType->isSuperTypeOf($type)->yes()) {
+                     return true;
+                 }
```

File: rules/NetteToSymfony/NodeAnalyzer/NetteControlFactoryInterfaceAnalyzer.php [hunk 1]
```diff
@@ -38,12 +34,16 @@ public function isComponentFactoryInterface(Interface_ $interface): bool
...
-            $className = $this->nodeTypeResolver->getFullyQualifiedClassName($returnType);
-            if (is_a($className, 'Nette\Application\UI\Control', true)) {
-                return true;
+            $controlOrForm = new UnionType([
+                new ObjectType('Nette\Application\UI\Control'),
+                new ObjectType('Nette\Application\UI\Form'),
+            ]);
+
+            if ($returnType instanceof ShortenedObjectType) {
+                $returnType = new ObjectType($returnType->getFullyQualifiedName());
```
Comment (TomasVotruba) [inline rules/NetteToSymfony/NodeAnalyzer/NetteControlFactoryInterfaceAnalyzer.php:43]: The check work without this wraparound, as in other places

Comment (samsonasik) [reply rules/NetteToSymfony/NodeAnalyzer/NetteControlFactoryInterfaceAnalyzer.php:43]: without this check, it got error in test like in my previous comment https://github.com/rectorphp/rector/pull/5907#issuecomment-802718573

**Comment by samsonasik** [comment]
@ondrejmirtes I tried update other places, but got errors like in my previous comments, I rollback to only apply what you suggested in #5906 .

**Comment by TomasVotruba** [comment]
I'll look into this

**Comment by TomasVotruba** [comment]
Thank you :+1:

**Pull request merged**